packages/apollo-federation/src/composition/validate/sdl/uniqueTypeNamesWithoutEnumsOrScalars.ts

```diff
@@ -30,25 +50,128 @@ export function UniqueTypeNamesWithoutEnumsOrScalars(
+     * 1) have the same kind (type, interface, input), extensions are excluded
+     * 2) are not entities
+     * 3) have the same set of fields
+     */
+    if (duplicateTypeNode) {
+      const possibleErrors: GraphQLError[] = [];
+      // By inspecting the diff, we can warn when field types mismatch.
+      // A diff entry will exist when a field exists on one type and not the other, or if there is a type mismatch on the field
+      // i.e. { sku: [Int, String!], color: [String] }
+      const { kind, fields } = diffTypeNodes(node, duplicateTypeNode);
+
+      const fieldsDiff = Object.entries(fields);
+      const typesHaveSameShape =
+        fieldsDiff.length === 0 ||
+        fieldsDiff.every(([fieldName, types]) => {
+          // If a diff entry has two types, then the field name matches but the types do not.
+          // In this case, we can push a useful error to hint to the user that we
+          // think they tried to define a value type, but one of the fields has a type mismatch.
+          if (types.length === 2) {
+            possibleErrors.push(
+              errorWithCode(
+                'VALUE_TYPE_FIELD_TYPE_MISMATCH',
+                `Found field type mismatch on expected value type. '${typeName}.${fieldName}' is defined as both a ${types[0]} and a ${types[1]}. In order to define '${typeName}' in multiple places, the fields and their types must be identical.`,
+                [node, duplicateTypeNode],
+              ),
+            );
+            return true;
+          }
+          return false;
+        });
+
+      // Once we determined that types have the same shape (name, kind, and field
+      // names), we can provide useful errors
+      if (typesHaveSameShape) {
+        // Report errors that were collected while determining the matching shape of the types
+        possibleErrors.forEach(error => context.reportError(error));
+
+        // Error if the kinds don't match
+        if (kind.length > 0) {
+          context.reportError(
+            errorWithCode(
+              'VALUE_TYPE_KIND_MISMATCH',
+              `Found kind mismatch on expected value type. '${typeName}' is defined as both a ${kind[0]} and a ${kind[1]}. In order to define ${typeName} in multiple places, the kinds must be identical.`,
+              [node, duplicateTypeNode],
+            ),
+          );
+        }
+
+        // Error if either is an entity
+        if (isTypeNodeAnEntity(node) || isTypeNodeAnEntity(duplicateTypeNode)) {
+          context.reportError(
+            errorWithCode(
+              'VALUE_TYPE_NO_ENTITY',
+              `Value types cannot be entities (using the @key directive). Please ensure that one type extends the other correctly, or remove the @key directive if this is not an entity.`,
+              [node, duplicateTypeNode],
+            ),
+          );
+        }
+
+        return false;
+      }
+    }
 
-    if (schema && schema.getType(typeName)) {
+    if (duplicateTypeNode) {
```

Comment:
This seems to be a continuation of the previous block (which is also using the same conditional `if (duplicateTypeNode)`), and could just be rolled into that?

Reply:
I'm glad you and @JakeDawkins pointed this out to me, since it was actually a _small_ bug resulting in not the exactly appropriate error message.

🎉